tests/rptest/tests/datalake/custom_partitioning_test.py

```diff
@@ -78,3 +91,124 @@ def test_configs(self):
+    ]
+}
+"""
+
+
+class DatalakeCustomPartitioningTest(RedpandaTest):
+    def __init__(self, test_ctx, *args, **kwargs):
+        super(DatalakeCustomPartitioningTest,
+              self).__init__(test_ctx,
+                             num_brokers=4,
+                             si_settings=SISettings(test_context=test_ctx),
+                             extra_rp_conf={
+                                 "iceberg_enabled": True,
+                                 "iceberg_catalog_commit_interval_ms": 5000,
+                             },
+                             schema_registry_config=SchemaRegistryConfig(),
+                             *args,
+                             **kwargs)
+
+    def setUp(self):
+        # redpanda will be started by DatalakeServices
+        pass
+
+    @cluster(num_nodes=6)
+    @matrix(cloud_storage_type=supported_storage_types(),
+            filesystem_catalog_mode=[False, True])
+    def test_basic(self, cloud_storage_type, filesystem_catalog_mode):
+        with DatalakeServices(self.test_context,
+                              redpanda=self.redpanda,
+                              filesystem_catalog_mode=filesystem_catalog_mode,
+                              include_query_engines=[QueryEngineType.SPARK
+                                                     ]) as dl:
+            # Create an iceberg topic with a custom partition spec and produce
+            # some data.
+
+            topic_name = "foo"
+            msg_count = 1000
+            partitions = 5
+            dl.create_iceberg_enabled_topic(
+                topic_name,
+                partitions=partitions,
+                iceberg_mode="value_schema_id_prefix",
+                config={
+                    "redpanda.iceberg.partition.spec":
+                    "(hour(redpanda.timestamp), event_type)"
+                })
+
+            value_serializer = AvroSerializer(
+                SchemaRegistryClient(
+                    {"url": self.redpanda.schema_reg().split(",")[0]}),
+                AVRO_SCHEMA_STR)
+            producer = SerializingProducer({
+                'bootstrap.servers':
+                self.redpanda.brokers(),
+                'key.serializer':
+                StringSerializer('utf_8'),
+                'value.serializer':
+                value_serializer,
+            })
+
+            timestamp = time.time()
+            for i in range(msg_count):
+                ev_type = random.choice(["type_A", "type_B"])
+                record = {
+                    "event_type": ev_type,
+                    "number": i,
+                    "timestamp_us": int(timestamp * 1000000),
+                }
+                producer.produce(
+                    topic=topic_name,
+                    # key to ensure that all partitions get some records
+                    key=str(uuid4()),
+                    value=record)
+            producer.flush()
```

Comment:
Is it possible for us to get very unlucky and for the records in this batch to also be in different hours?

Reply:
good catch, I was intending to have the same `timestamp_us` value for all records and partition based on that, but forgot to change the spec expression.